udffsck/main.c

```diff
@@ -14,13 +11,701 @@
+
+    if(path == NULL) {
+        err("No medium given. Use -h for help.\n");
+        exit(16);
+    }
+
+    if(blocksize > 0) {
+        force_sectorsize = 1;
+    }
+
+    msg("Medium to analyze: %s\n", path);
+
+    /* TODO remove. Replaced by check during opening
+    //Check if medium is mounted or not
+    FILE* mtab = setmntent("/proc/mounts", "r");
+    struct mntent* m;
+    struct mntent mnt;
+    char strings[4096];
+    while ((m = getmntent_r(mtab, &mnt, strings, sizeof(strings)))) {
+    dbg("%s\n", mnt.mnt_fsname);
+    if(strcmp(mnt.mnt_fsname, path) == 0) { //Match
+    err("Medium is mounted, therefore cannot be checked. Exiting.\n");
+    exit(16);
+    }
+    }
+
+    endmntent(mtab);
+    */
+
+    int flags = O_RDONLY;
+    // If is there some request for corrections, we need read/write access to medium
+    if(interactive || autofix) {
+        flags = O_RDWR;
+        dbg("RW\n");
+    }
+
+    if((fd = open(path, flags, 0660)) == -1) {
+        fatal("Error opening %s: %s.", path, strerror(errno));
+        exit(16);
+    } else {
+        int fd2;
+        int flags2;
+        char filename2[64];
+        const char *error;
+
+        if (fstat(fd, &stat) != 0) {
+            fatal("Cannot stat device '%s': %s\n", path, strerror(errno));
+            exit(16);
+        }
+
+        flags2 = flags; 
+        if (snprintf(filename2, sizeof(filename2), "/proc/self/fd/%d", fd) >= (int)sizeof(filename2))
+        {
+            fatal("Cannot open device '%s': %s\n", path, strerror(ENAMETOOLONG));
+            exit(16);
+        }
+
+        // Re-open block device with O_EXCL mode which fails when device is already mounted
+        if (S_ISBLK(stat.st_mode))
+            flags2 |= O_EXCL;
+
+        fd2 = open(filename2, flags2);
+        if (fd2 < 0)
+        {
+            if (errno != ENOENT)
+            {
+                error = (errno != EBUSY) ? strerror(errno) : "Device is mounted or mkudffs is already running";
+                fatal("Cannot open device '%s': %s\n", path, error);
+                exit(16);
+            }
+
+            // Fallback to orignal filename when /proc is not available, but this introduce race condition between stat and open
+            fd2 = open(path, flags2);
+            if (fd2 < 0)
+            {
+                error = (errno != EBUSY) ? strerror(errno) : "Device is mounted or mkudffs is already running";
+                fatal("Cannot open device '%s': %s\n", path, error);
+                exit(16);
+            }
+        }
+
+        close(fd);
+        fd = fd2;
+
+    }
+
+    if((fp = fopen(path, "r")) == NULL) {
+        fatal("Error opening %s: %s.", path, strerror(errno));
+        exit(16);
+    }
+    //Lock medium to ensure no-one is going to change during our operation. Make nonblocking, so it will fail when medium is already locked.
+    if(flock(fd, LOCK_EX | LOCK_NB)) { 
```

Comment:
I do not think that flock is a good idea. For block devices is for sure better O_EXCL.